ateam_common/src/udp_sender.cpp

```diff
@@ -0,0 +1,85 @@
+// Copyright 2021 A Team
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#include "ateam_common/udp_sender.hpp"
+
+#include <boost/bind.hpp>
+
+#include <algorithm>
+#include <string>
+
+namespace ateam_common
+{
+
+UDPSender::UDPSender(
+  std::string udp_address_string,
+  int16_t udp_port)
+: udp_socket_(io_service_),
+  receiver_endpoint_(
+    boost::asio::ip::make_address(udp_address_string),
+    udp_port)
+{
+  udp_socket_.open(receiver_endpoint_.protocol());
+
+  io_service_thread_ = std::thread(
+    [this]() {
+      io_service_.run();
+    });
+}
+
+UDPSender::~UDPSender()
+{
+  io_service_.stop();
+  if (io_service_thread_.joinable()) {
+    io_service_thread_.join();
+  }
+}
+
+void UDPSender::send(char * data, size_t length)
+{
+  if (length >= buffer_.size()) {
+    // RCLCPP_ERROR(get_logger(), "UDP send data length is larger than buffer");
```

Comment:
We probably should at least log a warning if this gets tripped.

Reply:
So I meant to ask about this, I was having problems getting this to actually compile. Is there something special I need on top of the roscpp header to be able to do Ros logs?

Reply:
Just tried again to get the details. It seems `get_logger()` requires a ROS2 node reference (or to be called within a class that inherits from ROS2 nodes). Since this is a vanilla C++ class, I don't see an easy way to get that without just passing down a node reference :/

Reply:
That's a good point, and printing is fine for now.
Another option for situations like this is for these kinds of helpers to pass error information back up to the caller so they can decide what to do with it.